libs/core/execution_base/include/hpx/execution_base/any_sender.hpp

```diff
@@ -0,0 +1,769 @@
+        {
+            return object ==
+                reinterpret_cast<base_type const*>(&embedded_storage);
+        }
+
+        bool empty() const noexcept
+        {
+            return get().empty();
+        }
+
+        void release()
+        {
+            HPX_ASSERT(!empty());
+
+            if (using_embedded_storage())
+            {
+                get().~base_type();
+            }
+            else
+            {
+                delete heap_storage;
+                heap_storage = nullptr;
+            }
+
+            object = get_empty_vtable<base_type>::call();
+        }
+
+    public:
+        sbo_storage()
+          : heap_storage(nullptr)
+          , object(get_empty_vtable<base_type>::call())
+        {
+        }
+
+        ~sbo_storage()
+        {
+            if (!empty())
+            {
+                release();
+            }
+        }
+
+        sbo_storage(sbo_storage&& other)
+          : heap_storage(nullptr)
+          , object(get_empty_vtable<base_type>::call())
+        {
+            *this = std::move(other);
+        }
+
+        sbo_storage& operator=(sbo_storage&& other)
+        {
+            if (&other != this)
+            {
+                if (!empty())
+                {
+                    release();
+                }
+
+                if (!other.empty())
+                {
+                    if (other.using_embedded_storage())
+                    {
+                        auto p =
+                            reinterpret_cast<base_type*>(&embedded_storage);
+                        auto otherp = reinterpret_cast<base_type*>(
+                            &other.embedded_storage);
+                        // We cast the pointers to void* to silence warnings
+                        // about copying a dynamic class (which don't apply
+                        // here)
+                        std::memcpy(static_cast<void*>(p),
+                            static_cast<void*>(otherp), embedded_storage_size);
+                        object = p;
+                    }
+                    else
+                    {
+                        heap_storage = other.heap_storage;
+                        other.heap_storage = nullptr;
+                        object = heap_storage;
+                    }
+
+                    other.object = get_empty_vtable<base_type>::call();
+                }
+            }
+
+            return *this;
+        }
+
+        base_type const& get() const noexcept
+        {
+            return *object;
+        }
+
+        base_type& get() noexcept
+        {
+            return *object;
+        }
+
+        template <typename Impl, typename... Ts>
+        void store(Ts&&... ts)
+        {
+            if (!empty())
+            {
+                release();
+            }
+
+            if constexpr (can_use_embedded_storage<Impl>())
+            {
+                base_type* p = reinterpret_cast<base_type*>(&embedded_storage);
```

Comment:
Should this be `Impl* p = ...; `?

Reply:
I'm not sure actually. Do you know if it makes a difference? `Impl *` is definitely not less correct at least, so I can change.

Reply:
The placement `new` expects a `void*`, so it probably doesn't make a difference in this case. It just looks more consistent to have the right pointer type. You probably would get away with even writing: `new (&embedded_storage) ...;`

Reply:
Unfortunately I can't use `&embedded_storage` directly, as it's a completely unrelated type (`aligned_storage`). I updated it to be `Impl*`.

Reply:
I was halfway joking anyways ;-)